src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> first;
+    private Node<T> last;
+    private static class Node<T>{
+        T item;
+        Node<T> prev;
+        Node<T> next;
+
+        Node(Node<T> prev, T item, Node<T> next){
+            this.prev = prev;
+            this.next = next;
+            this.item = item;
+        }
+    }
+
+    // Adds new value at the last position
     @Override
     public void add(T value) {
+        if (last != null) {
+            last.next = new Node<>(last, value, null);
+            last = last.next;
+        }else{
+            first = last = new Node<>(null, value, null);
+        }
+        size++;
     }
 
+    // Adds new value at the index position or if index is out of size,
+    // it throws IndexOutOfBoundsException
     @Override
     public void add(T value, int index) {
+        if (index >=0 && index <= size){
+            if (index == size) {
+                add(value);
+            }
+            else if(index == 0){
+                first = new Node<>(null, value, first);
+            }else{
+                Node<T> nextNode = getNode(index);
+                Node<T> prevNode = nextNode.prev;
+
+                prevNode.next = new Node<>(prevNode, value, nextNode);
+                nextNode.prev = prevNode.next;
+
+                size++;
+            }
+        }else{
+            throw new IndexOutOfBoundsException();
+        }
     }
 
+    // Extends LinkedList with list elements
     @Override
     public void addAll(List<T> list) {
+        for (T element : list){
+            add(element);
+        }
     }
 
+    // Returns an element value at the index position
     @Override
     public T get(int index) {
-        return null;
+        Node<T> temporaryNode = first;
+        int counter = 0;
+
+        while(temporaryNode != null && counter != index){
+            temporaryNode = temporaryNode.next;
+            counter++;
+        }
+        if (temporaryNode == null) throw new IndexOutOfBoundsException();
+        return temporaryNode.item;
+    }
+
+    // Returns an element at the index position
+    public Node<T> getNode(int index) {
+        Node<T> temporaryNode = first;
+        int counter = 0;
+
+        while(temporaryNode != null && counter != index){
+            temporaryNode = temporaryNode.next;
+            counter++;
+        }
+        if (temporaryNode == null) throw new IndexOutOfBoundsException();
+        return temporaryNode;
     }
```

Comment:
Isn't it better to iterate from the end if index is bigger than half of list size?